### DISCUSSION TITLE
Action object parameters

### DISCUSSION THREAD

**Discussion by e-cal** [AI-Planning/macq#55] [General]
In manually constructing traces, I noticed an issue similar the the fluent-value one with `Action.obj_params`.

The `Action` constructor takes a name and a list of `CustomObjects` which the action acts on. By storing this on the `Action` itself, it isn't clear which object it actually acted in that step because the list either holds all the objects it _could_ act on, or is overwritten each time the action occurs. 

Multiple actions with the same name ("pick up" acting on A, "pick up" acting on B) could be used but as with the fluents that doesn't seem right.

This one doesn't affect model extraction, but makes it difficult to print action information.

As it is, can't determine which block was actually "picked up" from within the `Action`'s `__str__` method unless multiple "pick up" actions exist:
```python

# test.py
pick_up = Action("pick up", obj_params = [blockA, blockB])

Trace(
  [
    Step(
      State(
        {
           a_clear: True,
           b_clear: True,
           a_on_table: True,
           b_on_table: True,
         }
       ),
         pick_up,
       ), 
       ...
  ]
)

# action.py
def __str__(self):
        string = f"{self.name} {' '.join(map(str, self.obj_params))}"
        return string
```

Output of printing the trace list:
![image](https://user-images.githubusercontent.com/47398876/119841889-3404b500-bed4-11eb-9254-3ba4837358dd.png)

I think the solution would be to store action parameters in `Step`, 
ie `Action(name)` and `Step(state, action, objs)` 
instead of `Action(name, objs)` and `Step(state, action)` as it is now.

But seeing as this doesn't pose a problem to model acquisition, maybe it is better to leave it and just leave out `obj_params` in the `Action` string?

**Comment by haz** [comment]
Not sure what the core problem is -- a ground action is one that combines the lifted action (identifiable by name) and the object parameters. I think a step makes sense still as a tuple of state / ground-action. The objects don't really relate to the state (at least not directly).

Is the issue that an action object is uniquely defined (`__hash__`) using the action name?

**Reply by e-cal** [reply]
Oh I see, so should our `Action` class should be a grounded action then? and have separate actions for "pick up A" vs "pick up B"?

In my example "pick up A" isn't differentiable from "pick up B", because I just made one action named "pick up". I was thinking of our `Action` as similar to how in pddl you'd have something like 
```lisp
(:action pickup
    :parameters (?ob)
...)
```

**Reply by haz** [reply]
Well the longer-term view is we should have both! For now, the grounded action is the important one, and we should be able to distinguish between the `pickup`'s.

A broader question (and one @beckydvn should step in on), is if/what we need to represent explicitly -vs- rely on `tarsk` for.

**Reply by beckydvn** [reply]
"pick up A" and "pick up B" would be separate/distinguishable Action objects. Not only do they act on different objects but during parsing, a new `Action` is created every `Step`. 
You could end up having "duplicate" actions on different steps; even actions with the same name that act on objects with the same name (but they are still different objects). If you wanted somehow to differentiate between the "pick up A" Action on step 2 versus "pick up A" Action on step 6 for example, you would have to do that by comparing the objects themselves, because otherwise the attributes would be the same. I'm not sure in what case you would want/need to do that, though(?)

**Reply by e-cal** [reply]
Oh, I didn't realize each step had unique actions!

So if there are a bunch of unique instance, should `extract` combine all the identical actions into one by cross referencing all actions with the same attributes (use `__eq__`), or should it keep all the individual instances (use `is`)?

If it should combine, then I need this in order to find the `add`, `delete`, and `pre_cond` for the action. The `extract` method will need to cross reference the sas triples for each instance of the same action (or `get_sas_triples` will need to compare to get lists >1 element)

**Reply by beckydvn** [reply]
What do you mean by "combine all the identical actions into one"? I think I'm missing something here, but just because actions are "identical" doesn't mean the information surrounding them is, so I don't understand why they would be combined. i.e. "pick up A" on step 3 may have a different state before it than "pick up A" on step 11. In which case, you would need to examine the sas triples for every instance.

**Reply by beckydvn** [reply]
Even if the actions in the trace appear "identical", having the same name and acting on objects with the same name, I feel like you can't really treat them as identical because the information around them is different. I think I may just be misunderstanding your question though. What do you think?

**Reply by e-cal** [reply]
I mean should the `Model`'s list of actions have ["pick up A", "pick up A", "pick up A", ...] or just one "pick up A"?

**Reply by beckydvn** [reply]
I see... is the Model storing grounded actions? I thought it would just be storing actions in a general sense (i.e. pickup an object of type block).

**Reply by beckydvn** [reply]
In which case, pick up would just be one action.

**Reply by beckydvn** [reply]
I suppose if you wanted to have both grounded and non-grounded actions available, then yes it makes sense to combine them and just have one "pick up A." I thought you were referring to combining them in the analysis stage, my bad :)

**Reply by haz** [reply]
Sorry I'm only getting to this conversation now :P.

We will eventually want the option for both. There may be cases down the road when multiple actions with the same name and parameters need to be kept separate during extraction because of certain types of execution uncertainty.

For now, I think it'd be ok to define the `__eq__` (or at least the `__hash__`) as corresponding to the action+parameters. This will help in things like "give me all the triples with this ground action". When it comes time for the more sophisticated representation, we can subclass.

From the Python perspective, there may be multiple objects for `(pickup A)`, but with the right `__hash__` implementation, they'd only appear once in sets we care to create, comparisons we make, etc.

**Reply by e-cal** [reply]
Ok perfect, thanks for the advice guys!
